jbpm-runtime-manager/src/main/java/org/jbpm/runtime/manager/impl/PerProcessInstanceRuntimeManager.java

```diff
@@ -191,15 +194,35 @@ public void signalEvent(String type, Object event) {
         
         // process currently active runtime engines
         Map<Object, RuntimeEngine> currentlyActive = local.get();
+
         if (currentlyActive != null && !currentlyActive.isEmpty()) {
-            RuntimeEngine[] activeEngines = currentlyActive.values().toArray(new RuntimeEngine[currentlyActive.size()]);
-            for (RuntimeEngine engine : activeEngines) {
-                Context<?> context = ((RuntimeEngineImpl) engine).getContext();
+            @SuppressWarnings("unchecked")
+            Entry<Object, RuntimeEngine> activeEngines[] = currentlyActive.entrySet()
+                    .toArray(new Entry[currentlyActive.size()]);
+            Set<Object> enginesToDelete = new HashSet<>();
+            for (Entry<Object, RuntimeEngine> engine : activeEngines) {
+                RuntimeEngineImpl engineImpl = (RuntimeEngineImpl) engine.getValue();
+                if (engineImpl.isDisposed() || engineImpl.isInvalid()) {
```

Comment:
add a not-null check to skip possible null pointer
```suggestion
                if (engineImpl==null) {
                     continue;
                }
                if (engineImpl.isDisposed() || engineImpl.isInvalid()) {
```